Initial implementation of accessing external probes

fbertsch opened the pull request:
These changes are meant to add other repositories and their
probes to the probe-scraper output.

The users input their repo information in repositories.yaml.
The program is meant to run daily, and each day the scraped files
are added to the previously scraped ones. This means that the
previous day's scrapes need to be available, which can be done by
having a dedicated machine or by using external storage and syncing.

If there is a problem downloading the file or processing it,
an email is sent with the issue.

The probes are outputted in similar format, except that they
do not have _version_ or _node_. Instead they have _date_,
which contains the first and last date we saw that definition.

The files are outputted to
`output_dir/<repo_name>/mobile-metrics/all-probes`.

Eventually we can include whether it should be `main` ping probes
(which indicates addon scalars and histograms) or mobile-metrics,
so this is extensible to future data.

*Summary of Changes*
- Move scrapers to `/scrapers/`, rename existing scraper to
  `moz_central_scraper`
- Add `repositories.yaml`, which indicates external repos to
  scrape
- Add `http_scraper`, which scrapes all files in `repositories.yaml`
- Add separate runner for scraping and storing http-accessed repos
- Add `emailer.py`, which is a simple email API
- Add capability to `transform.py` to merge definitions by date
- Add tests to `transform.py` for this capability

georgf commented on probe_scraper/runner.py:
```diff
@@ -102,19 +120,84 @@ def main(temp_dir, out_dir):
+    #   ...
+    # }
+    probes = defaultdict(lambda: defaultdict(lambda: defaultdict(int)))
+    for repo_name, dates in repos_probes_data.iteritems():
```
AFAICT accessing this loaded structure has multiple points where it can fail due to invalid structure.
In other places we use a pattern of "load", "check structure", "process data", where "check structure" does a basic schema check.
Can we add a simple schema check here to avoid erroring out unexpectedly or adding lots of detailed error handling?

georgf commented on probe_scraper/runner.py:
```diff
@@ -102,19 +120,84 @@ def main(temp_dir, out_dir):
                         help='Directory to store output files in.',
                         action='store',
                         default='.')
+    parser.add_argument('--skip-moz-central-probes',
```
This might be easier to understand as `--only-moz-central-probes` or similar.
This is not a blocker if that's involved to change.

georgf commented on probe_scraper/runner.py:
```diff
@@ -102,19 +120,84 @@ def main(temp_dir, out_dir):
                         action='store',
                         default='.')
+    parser.add_argument('--skip-moz-central-probes',
+                        help='Directory to store output files in.',
```
Update the help here and below.

georgf commented on probe_scraper/runner.py:
```diff
@@ -102,19 +120,84 @@ def main(temp_dir, out_dir):
-    parser.add_argument('--tempdir',
-                        help='Temporary directory to work in. This serves as a cache if reused.',
+    parser.add_argument('--date',
+                        help='The date being run',
```
What is this used for? Can you expand the help?

georgf commented:
On the output path format, `output_dir/<repo_name>/mobile-metrics/all_probes`
... are we expecting to explicitly go for each individual mobile repos probes?
Or would we mostly access this as "give me all the mobile-metrics probes for all mobile projects"?

georgf commented:
Also, could you put example output from this change somewhere that we can look at?

Dexterp37 commented on probe_scraper/runner.py:
```diff
@@ -39,40 +45,52 @@ def general_data():
+    dump_json(revisions, base_dir, "revisions")
 
-    # Break down the output by channel. We don't need to write a revisions
+    # Break down the output by channel. We don"t need to write a revisions
```
nit: this should be `don't` not `don"t` :)

fbertsch commented:
@georgf RFAL. Output is available at https://s3-us-west-2.amazonaws.com/telemetry-public-analysis-2/frank/mobile_metrics_example/mobile-metrics/all_probes.json

fbertsch commented:
@georgf to make this a bit easier, you can just pull down my branch and run `python probe_scraper/runner.py --only-git-probes` and it will output the correct files.

Here's a quick outline of what's happening in the code:

1. `git_parser.py`
- parses the repositories registry file
- for each repository, clone it
  - for each probe file, get all of the commits that touched it, and retrieve the file for that commit
2. `runner.py`
- take an input of per-commit probe files, and output a list of every commit and every file within that commit
3. `transform_probes.py`
- start, from the first commit (ordered by timestamp of the commit) a list of all probes
- if probes are the same on the next commit, update the `commits.last` element for the definition

The only other additions have been:
- A schema for the interim data structure outputted from `git_parse.py`

I do think the emailer is good as-is, because most of the errors it catches should be issues with the files themselves, not code issues. Let me know if you find that's not the case.

georgf commented on probe_scraper/parsers/scalars.py:
```diff
@@ -7,11 +7,18 @@
+    try:
+        optout = s.dataset.endswith('_OPTOUT')
+    except KeyError:
+        optout = s._definition.get('release_channel_collection', 'prerelease') == 'prerelease'
```
Per [the design doc](https://docs.google.com/document/d/10FVEYIFkybRotbe30FquAGE7O0K92bAKSinvrsCfh6o/edit?ts=5af99e01#heading=h.agjq0277ox7i) this should be `collect_on_channels`?

Can we just add a property in `parse_scalars.py` that encapsulates this behavior?

fbertsch replied:
I agree, it should be part of `parse_scalars.py`. How about we keep as-is, and I'll make a bug for adding an `is_optout()` function to `parse_scalars.py`, and when that's merged we can update this?

georgf commented on probe_scraper/runner.py:
```diff
@@ -2,28 +2,35 @@
 
 
+FROM_EMAIL = "telemetry-alerts@mozilla.com"
+DEFAULT_TO_EMAIL = "dev-telemetry-alerts@mozilla.com"
```
Hm, is that a good default email list? Do we have anything more suitable?
This would overload that list usage even more.

fbertsch replied:
I filter those emails pretty aggressively, if people want to ignore them they can always just send `subject contains "Probe Scraper"` to trash. I'm not sure of a better alternative, did you have any ideas?

georgf replied:
Right, that's maybe better as a future point.
Can you give Chris a heads-up on this?

fbertsch commented on README.md:
```diff
@@ -143,6 +162,72 @@ This file contains the data for the probes. The data might be spread across mult
+          },
+          "expiry_version": "<string>",
+          "optout": <bool>,
+          "commits": {
```
Yup, that's a good idea.

fbertsch commented on README.md:
```diff
@@ -7,6 +7,23 @@ Also, probes outside of Histograms.json - like the CSS use counters - are includ
+will be just about your specific repository.
+- `url`: The URL of the repository to scrape. It should be able to be cloned directly from that URL.
+- `histogram`: A list of relative paths to `Histograms.json` files
+- `scalar`: A list of relative paths to `Scalars.yaml` files
```
👍

georgf commented on probe_scraper/emailer.py:
```diff
@@ -0,0 +1,62 @@
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import boto3
```
Is this missing from requirements.txt?

fbertsch replied:
Added.

georgf commented on probe_scraper/parsers/scalars.py:
```diff
@@ -7,11 +7,18 @@
+    try:
+        optout = s.dataset.endswith('_OPTOUT')
+    except KeyError:
+        optout = s._definition.get('collect_on_channels', 'prerelease') == 'release'
```
Why not add this as a property to the parser that encapsulates this?
(Hm, did i comment on this before?)

fbertsch replied:
You did, this would be a dependency on `parse_scalars.py`.

georgf commented on repositories.yaml:
```diff
@@ -0,0 +1,10 @@
+    - frank@mozilla.com
+  url: 'https://github.com/fbertsch/mobile-metrics-example'
+  histogram_file_paths:
+    - Histograms.json
```
Why not parse the Histograms.yaml?
Histograms.json has the old format, with no categories. Can we just not support that?

fbertsch replied:
I thought we had discussed supporting both, but extend this later to include `Histograms.yaml`. There's a dependency on the `Histograms` class in `python_moztelemetry`. Let me see how much work that might be.

fbertsch commented on probe_scraper/runner.py:
```diff
@@ -102,19 +121,96 @@ def main(temp_dir, out_dir):
+    schema = Schema({
+        str: {
+            And(str, lambda x: len(x) == 40): {
+                Optional("histogram"): [And(str, lambda x: os.path.exists(x))],
```
I wanted this to be as close to the mozilla-central data structure as possible, which uses `histogram`, `scalar`, etc.

fbertsch commented on probe_scraper/runner.py:
```diff
@@ -102,19 +121,96 @@ def main(temp_dir, out_dir):
+    group = parser.add_mutually_exclusive_group()
+    group.add_argument('--only-moz-central-probes',
+                       help='Only scrape moz-central probes',
+                       action='store_true')
```
`store_true` implies default false, so I added Alessio's comment and things should be good to go

fbertsch commented on probe_scraper/transform_probes.py:
```diff
@@ -174,3 +186,110 @@ def transform(probe_data, node_data, break_by_channel):
+    """
+
+    all_probes = defaultdict(lambda: defaultdict(lambda: defaultdict(list)))
+    for repo_name, commits in probe_data.iteritems():
```
Updated.

fbertsch commented:
@Dexterp37 okay, it's ready for a final-pass. I added integration tests with a auto-created repos. I also split up some of the `transform_probes.py` code to make it more human-readable.

mreid-moz commented on probe_scraper/transform_probes.py:
```diff
@@ -174,3 +186,128 @@ def transform(probe_data, node_data, break_by_channel):
+                [new_defn] + prev_defns
+
+    # We haven't seen this probe before, add it
+    else:
```
qq, which style guide are you referring to?

mreid-moz replied:
Nice, thanks for the link. I'll add it to the team Onboarding materials.

Dexterp37 replied:
Fine by me! Let's not block on this (and the similar ones below)!

fbertsch commented on probe_scraper/transform_probes.py:
```diff
@@ -174,3 +186,128 @@ def transform(probe_data, node_data, break_by_channel):
+            ]
+
+            for probe, probe_type, definition in probes:
+                all_probes = update_or_add_probe(all_probes,
```
No, but generally I prefer a functional approach because it makes things more obvious.

fbertsch commented:
@Dexterp37 The job successfully ran on an ATMO cluster
